Proxied iter optimizations

klaussfreire opened the pull request:
- Support iteration modes with reusable proxies
- Optimize iteration factoring metadata querying out of the loops

Timings (l = python list, c = proxied_list):

Patched:
``` pycon
>>> timeit.timeit(lambda : 3 in l, number=1000000)
1.8466320037841797
>>> timeit.timeit(lambda : 3 in c, number=1000000)
23.526846170425415
>>> timeit.timeit(lambda : 3 in list(c), number=1000000)
33.05595397949219
>>> timeit.timeit(lambda : list(l), number=1000000)
0.37212204933166504
>>> timeit.timeit(lambda : list(c), number=1000000)
31.120682954788208
>>> timeit.timeit(lambda : list(iter(l)), number=1000000)
1.1441419124603271
>>> timeit.timeit(lambda : list(iter(c)), number=1000000)
33.150113105773926
>>> timeit.timeit(lambda : list(c.iter()), number=1000000)
31.735125064849854
>>> timeit.timeit(lambda : list(c.iter_fast()), number=1000000)
21.98051691055298
>>> timeit.timeit(lambda : list(c.iter()), number=1000000)
32.060221910476685
>>> timeit.timeit(lambda : map(l.__getitem__, xrange(len(c))), number=1000000)
2.8401761054992676
>>> timeit.timeit(lambda : map(c.__getitem__, xrange(len(c))), number=1000000)
33.33895683288574
>>> timeit.timeit(lambda : map(c.getter(), xrange(len(c))), number=1000000)
32.38253593444824
>>> timeit.timeit(lambda : map(c.getter(fast=True), xrange(len(c))), number=1000000)
24.03903603553772
```

Unpatched:
``` pycon
>>> timeit.timeit(lambda : 3 in l, number=1000000)
1.6442980766296387
>>> timeit.timeit(lambda : 3 in c, number=1000000)
34.34405493736267
>>> timeit.timeit(lambda : 3 in list(c), number=1000000)
36.01886010169983
>>> timeit.timeit(lambda : list(l), number=1000000)
0.3630411624908447
>>> timeit.timeit(lambda : list(c), number=1000000)
33.33185315132141
>>> timeit.timeit(lambda : list(iter(l)), number=1000000)
1.0404939651489258
>>> timeit.timeit(lambda : list(iter(c)), number=1000000)
34.80966806411743
>>> timeit.timeit(lambda : map(l.__getitem__, xrange(len(c))), number=1000000)
2.8532519340515137
>>> timeit.timeit(lambda : map(c.__getitem__, xrange(len(c))), number=1000000)
35.17455792427063
```

klaussfreire committed:
Optimize proxied_list iteration

Factor _metadata out of the loop

klaussfreire committed:
Fast iteration interface that allows proxy reuse

klaussfreire committed:
Fast getter factory for proxied_list

llogiudice commented on sharedbuffers/mapped_struct.py:
```diff
@@ -1408,15 +1408,37 @@ def unpack_from(cls, buf, offs, idmap = None):
+    @cython.locals(fast = cython.bint, dcode = cython.char, index = cython.longlong,
+        objlen = cython.longlong, dataoffs = cython.Py_ssize_t, itemsize = cython.uchar)
+    def getter(self, proxy_into = None, fast = False):
+        if proxy_into is None and fast:
```
I think it would be worth it to delay constructing a generic proxy until you know that the datatype of the list is not purely numeric (i.e: dcode equal to 't' or 'T'). This way, you're eagerly creating a new object, even though in many cases it won't be needed.

klaussfreire replied:
Good point

klaussfreire committed:
Avoid creating a proxy in numeric lists

klaussfreire committed:
Add changelog entry

klaussfreire committed:
Small optimization in proxy reuse

Short-circuit to avoid calling PyObject_IsInstance unnecessarily

klaussfreire committed:
Add missing cclass directives

klaussfreire committed:
Fix pure-python

klaussfreire committed:
Prefer packing sets of signed ints

As unsigned ones produce long ints instead of regular ints,
which are bulkier and slower in python 2.

klaussfreire committed:
Avoid unnecessary temporary lists

klaussfreire committed:
More accurate comment

klaussfreire committed:
Lower-overhead super calls

klaussfreire committed:
Various improvements:

- Masked iteration in iter_fast
- Simplified NuericIdMultiMapper subclassing

klaussfreire committed:
Masked iteration for non-fast iter

klaussfreire committed:
Test masked iteration

Switch to uint8 arrays (Cython's typed memoryview support doesn't
include support for np bools)

klaussfreire committed:
Various object unpacking optimizations

Object unpacking is the bottleneck during proxied_list iteration

klaussfreire committed:
Favor regular int for small bitmaps in unpacker caches

klaussfreire committed:
Favor small int keys in packer cache

klaussfreire committed:
Add another cache layer to get_unpacker

Use a regular dict, that can be accessed efficiently from cython

klaussfreire committed:
Fix cython cast of long literal

klaussfreire committed:
mapped_list.unpack_from optimizations

klaussfreire committed:
We now require Cython 0.29

Also minor optimization
